gnocchi/chef.py

```diff
@@ -54,6 +54,44 @@ def __init__(self, coord, incoming, index, storage):
         self.index = index
         self.storage = storage
 
+    def auto_clean_expired_resources(self, resource_ended_at_normalization):
+        """Cleans expired resources.
+
+        This method will clean resources that have expired according to their 'ended_at' field. The method itself will
+        not execute the cleanup, we will mark the resource as deleted, and leave for the system to execute the actual
+        removal of the data.
+        """
+        moment_now = utils.utcnow()
+        moment = moment_now - datetime.timedelta(seconds=resource_ended_at_normalization)
+        attribute_filter = {"<": {"ended_at": moment}}
+
+        all_resources_found = self.index.list_resources(attribute_filter=attribute_filter)
+        LOG.debug("Resources found for deletion: [%s] that have been expired since [%s].", all_resources_found, moment)
+
+        auto_clean_lock = self.get_sack_lock("auto_clean_expired_resources_lock")
```

Comment:
not sure exactly the size of the resources table and if the "ended_at" column has or not an index, but I guess is cheaper to require a lock than run the list resources query, then maybe getting the lock as the first thing you do in this cleanup process, would save some DB processing. In this case, the try / finally should be for the entire method.

Reply:
the ended_at has index. I also put everything inside the try/finally after the lock.